### DISCUSSION TITLE
CSS/images of a template not loading

### DISCUSSION THREAD

**Discussion by Shahin-rmz** [zauberzeug/nicegui#867] [Q&A] [answered]
### Question

This is gonna be probably a basic question.
I can load a html template with Jinja.
It works well.
here is how it works:
```python
@app.get("/invoice/{invoice_id}", response_class=HTMLResponse)
def invoice( request: Request):
    subtotal = sum(row['sum_unit'] for row in rows)
    userid2 = id_of_invoice()
    return templates.TemplateResponse("final_test2.html",{"request": request,
                                                         "rows":rows,
                                                         "subtotal":subtotal,
                                                         "date":date,
                                                         "invoice_id": userid2})
```
and also 
```python
ui.button("make an invoice",on_click=lambda:(
          ui.open(f'/invoice/{invoice_id}'))).classes('''bg-yellow
                                        text-black''')
```

So far so good.
The problem is with loading CSS file and Images.

I have given the address of css inside the html file, and when I open it in the browser it works without problem.
Here is how it works.
```html
<DOCTYPE html>
  <html>
    <head>
      <title>My Items</title>
      <link rel="stylesheet" href="./style.css" />
    </head>
    <body>
      <div class="main-wrapper">
        <div class="inner-wrapper">
          <img src="./img/header.PNG" alt="header" />
          <div class="content-wrapper">
            <div class="header">
              <div class="inner-header">
                <h1>INVOICE</h1>
.
.
.
```
But when I open the template, I get the following error.

```
NiceGUI ready to go on http://0.0.0.0:8080
http://127.0.0.1:8080/invoice/img/header.PNG not found
http://127.0.0.1:8080/invoice/img/phone-call-svgrepo-com.svg not found
http://127.0.0.1:8080/invoice/img/email-8-svgrepo-com.svg not found
http://127.0.0.1:8080/invoice/img/footer.PNG not found
http://127.0.0.1:8080/invoice/img/header.PNG not found
http://127.0.0.1:8080/invoice/img/phone-call-svgrepo-com.svg not found
http://127.0.0.1:8080/invoice/img/email-8-svgrepo-com.svg not found
http://127.0.0.1:8080/invoice/img/footer.PNG not found

```
Also tried to serve css and images inside my app.
here is the code
```python
app.add_static_files("/templates","templates")
app.add_static_files("/templates/img","img")
```
and here is the tree of the directory tree.
```
├── main.py
├── grid.py
├── img
│   ├── email-8-svgrepo-com.svg
│   ├── footer.PNG
│   ├── header.PNG
│   └── phone-call-svgrepo-com.svg
├── jinja
│   ├── bin
│   ├── etc
│   ├── include
│   ├── lib
│   ├── lib64 -> lib
│   ├── pyvenv.cfg
│   └── share
├── templates
│   ├── d.zip
│   ├── final_test2.html
│   ├── greeting.html
│   ├── img
│   └── style.css
├── test_for_loop.py
└── test.py
```

also here are the permissions:
```
drwxrwxr-x 7 shahin shahin 4,0K Apr 25 09:34 jinja
-rw-rw-r-- 1 shahin shahin  702 Apr 27 13:21 test.py
-rw-rw-r-- 1 shahin shahin  589 Apr 27 14:19 grid.py
-rw-rw-r-- 1 shahin shahin  888 Apr 27 14:21 test_for_loop.py
drwxr-xr-x 2 shahin shahin 4,0K Apr 29 08:27 img
drwxrwxr-x 3 shahin shahin 4,0K Apr 30 18:38 templates
-rw-rw-r-- 1 shahin shahin 3,3K Apr 30 19:25 main.py

```

**Comment by rodja** [comment, marked as answer]
You have added the static files at `/templates` but you probably use a relative path starting with `./` in your html file. Change it to `/templates/img/header.png` or `/img/header.png` depending on you have used in `ui.add_static_files`.

**Comment by Shahin-rmz** [comment]
Thanks,
yes, That was the solution.
This is what I have been end up with.
```
app.add_static_files("/home/....../jinja_test/templates/","templates")
app.add_static_files("/home/......jinja_test/templates/d/img","img")
```

and also in all html files:

```html
<link rel="stylesheet" href="/home/..../jinja_test/templates/d/style.css" />
```
This works fine.
But I am not sure why we can not use relative path in html or our app.

Thanks

**Reply by rodja** [reply]
You can use relative paths. But they need to be correct. If your html is served at `/invoice/{invoice_id}` and you have a `src="./img/header.png"` in that html, the header must be available to the browser as `/invoice/img/header.png`, which may not be what you want. If you make the images available at `/templates/img` the src can be changed to `../templates/img/header.png` if you want relative paths. Or just use `src="/templates/img/header.png"`. The absolute paths starting with `/home` is very unconventional (even if it may work). The url paths have nothing to do with the file locations on disk but rather represent the resource paths of your web app.

**Reply by Shahin-rmz** [reply]
> You can use relative paths. But they need to be correct. If your html is served at `/invoice/{invoice_id}` and you have a `src="./img/header.png"` in that html, the header must be available to the browser as `/invoice/img/header.png`, which may not be what you want. If you make the images available at `/templates/img` the src can be changed to `../templates/img/header.png` if you want relative paths. Or just use `src="/templates/img/header.png"`. The absolute paths starting with `/home` is very unconventional (even if it may work). The url paths have nothing to do with the file locations on disk but rather represent the resource paths of your web app.

Thanks for informative and complete answer. My apology. The mistake was because of my lack of knowledge in Front end.
Best
